cpp/include/cuspatial/experimental/detail/point_linestring_distance.cuh

```diff
@@ -41,176 +41,80 @@ namespace cuspatial {
+    // Search from the linestring geometry offsets array to determine the geometry idx of
+    // current linestring point
+    auto geometry_idx = multilinestrings.geometry_idx_from_part_idx(part_idx);
 
     // Reduce the minimum distance between different parts of the multi-point.
-    vec_2d<T> const a      = linestring_points_first[idx];
-    vec_2d<T> const b      = linestring_points_first[idx + 1];
+    auto [a, b]            = multilinestrings.segment(idx);
     T min_distance_squared = std::numeric_limits<T>::max();
 
-    for (auto point_idx = point_geometry_offset_first[geometry_idx];
-         point_idx < point_geometry_offset_first[geometry_idx + 1];
-         point_idx++) {
-      vec_2d<T> const c = points_first[point_idx];
-
+    for (vec_2d<T> const& c : multipoints[geometry_idx]) {
       // TODO: reduce redundant computation only related to `a`, `b` in this helper.
       auto const distance_squared = point_to_segment_distance_squared(c, a, b);
-      min_distance_squared        = std::min(distance_squared, min_distance_squared);
+      min_distance_squared        = min(distance_squared, min_distance_squared);
     }
-
     atomicMin(&thrust::raw_reference_cast(*(distances + geometry_idx)),
-              static_cast<T>(std::sqrt(min_distance_squared)));
+              static_cast<T>(sqrt(min_distance_squared)));
   }
 }
 
 }  // namespace detail
-
-template <class Cart2dItA,
-          class Cart2dItB,
-          class OffsetIteratorA,
-          class OffsetIteratorB,
-          class OffsetIteratorC,
-          class OutputIt>
-OutputIt pairwise_point_linestring_distance(OffsetIteratorA point_geometry_offset_first,
-                                            OffsetIteratorA point_geometry_offset_last,
-                                            Cart2dItA points_first,
-                                            Cart2dItA points_last,
-                                            OffsetIteratorB linestring_geometry_offset_first,
-                                            OffsetIteratorC linestring_part_offsets_first,
-                                            OffsetIteratorC linestring_part_offsets_last,
-                                            Cart2dItB linestring_points_first,
-                                            Cart2dItB linestring_points_last,
+template <class MultiPointRange, class MultiLinestringRange, class OutputIt>
+OutputIt pairwise_point_linestring_distance(MultiPointRange multipoints,
+                                            MultiLinestringRange multilinestrings,
                                             OutputIt distances_first,
                                             rmm::cuda_stream_view stream)
 {
-  using T = iterator_vec_base_type<Cart2dItA>;
+  using T = typename MultiPointRange::element_t;
 
-  static_assert(is_same_floating_point<T, iterator_vec_base_type<Cart2dItB>>(),
+  static_assert(is_same_floating_point<T, typename MultiLinestringRange::element_t>(),
```

Comment:
Isn't this assertion covered by the `is_same` assertion below? If the vec_2ds are the same type, then the T must be the same as well.

Reply:
The one below doesn't test floating points type.